### DISCUSSION TITLE
adl_serializer specialization fails with GCC / works with Clang?

### DISCUSSION THREAD

**Discussion by lukasberbuer** [nlohmann/json#3166] [Q&A] [answered]
I try to implement an `enum` serialization with `magic_enum` based on a partial specialization of `nlohmann::adl_serializer`.
Following implementation **works with Clang** but **throws and compile error with GCC**:

```
error: partial specialization of 'struct nlohmann::adl_serializer<T, typename std::enable_if<is_enum_v<T>, void>::type>' after instantiation of 'struct nlohmann::adl_serializer<nlohmann::detail::value_t, void>' [-fpermissive]
```

Minimal example:
```cpp
#include <iostream>
#include <type_traits>

#include <nlohmann/json.hpp>
#include <magic_enum.hpp>

namespace nlohmann {

template <typename T>
struct adl_serializer<T, std::enable_if_t<std::is_enum_v<T>>> {
    static void to_json(json& j, const T& e) {
        j = magic_enum::enum_name(e);
    }

    static void from_json(const json& j, T& e) {
        const auto opt = magic_enum::enum_cast<T>(j.get<std::string_view>());
        e = opt.value_or(T{});  // use first enum value as fallback
    }
};

}  // namespace nlohmann

enum class Enum { test, x, y, z };

int main() {
    nlohmann::json j = Enum::y;

    // serialize
    std::cout << j.dump() << std::endl;

    // parse
    const auto e = j.get<Enum>();
    std::cout << static_cast<int>(e) << std::endl;
}
```

Link to godbolt / GCC: https://godbolt.org/z/4G6vPhKMe
Link to godbolt / Clang: https://godbolt.org/z/PobWEnj7G

How can I get it work in both compilers?

**Comment by gregmarr** [comment, marked as answer]
Looks like the `adl_serializer` is used internally for `nlohmann::detail::value_t` which is an enum, so it's an error for you to try to create a class that handles all enums.  This would also fail if you tried to use it in any other program that implements `adl_serializer` for any enum.  I suspect that you will need to opt-in to this function on a per-enum basis via some method.

**Reply by lukasberbuer** [reply]
Thanks! Adding the condition `std::enable_if_t<std::is_enum_v<T> && !std::is_same_v<T, detail::value_t>>` fixes that issue. But I get your point that a opt-in is the safer choice.
